functions/src/aqi-calculation/purple-air-response.ts

```diff
@@ -1,32 +1,370 @@
+    headers: {
+      'X-API-Key': config.purpleair.read_key,
+    },
+    params: {
+      fields: fieldList.join(),
+      max_age: maxSensorAge, // eslint-disable-line camelcase
+    },
+  });
+
+  return purpleAirResponse;
+}
+
+/**
+ * Converts PurpleAir's confidence value into percent difference
+ * @param confidence - confidence value from PurpleAir, between 0 and 100
+ * @returns meanPercentDifference value, or NaN if value lost in calculation.
+ * Any input that results in NaN means that the meanPercentDifference is high
+ * enough that the reading should be discarded anyways.
+ *
+ * @remarks
+ * PurpleAir's confidence values is calculated as follow. PurpleAir does not
+ * document what `pc` means.
+ * ```ts
+ * // a is the pseudo average for channel A
+ * // b is the pseudo average for channel B
+ * function getConfidence(a: number, b: number) {
+ *   const diff = Math.abs(a - b)
+ *   const avg = (a + b) / 2;
+ *   const meanPercentDifference = (diff / avg) * 100;
+ *   const pc = Math.max(Math.round(meanPercentDifference / 1.6) - 25, 0);
+ *   return Math.max(100 - pc, 0);
+ * }
+ * ```
+ */
+function getMeanPercentDifference(confidence: number): number {
+  const minConfidence = 0;
+  const maxConfidence = 100;
+  /* eslint-disable no-magic-numbers */
+  switch (confidence) {
+    case minConfidence:
+      // If the confidence is zero, then we want to discard this reading
+      return NaN;
+    case maxConfidence:
+      // The confidence value from PurpleAir can be 100 even if channel A and
+      // channel B do not completely match, but if the confidence value is 100,
+      // then the value is data is good enough to meet the EPA recommendation.
+      return 0;
+    default:
+      // Otherwise, undo the calculation from the PurpleAir confidence value
+      /* eslint-disable-next-line no-magic-numbers */
+      return ((maxConfidence - confidence + 25) * 1.6) / 100;
+  }
+}
+
+/**
+ * Converts a PurpleAir reading returned from the group API query into a PurpleAirReading
+ * @param data - list of data from PurpleAir for a given sensor
+ * @param fieldNames - list of field names from PurpleAir that match the order of the data fields
+ * @returns
+ */
+function getReading(
+  data: (string | number)[],
+  fieldNames: string[]
+): [number, PurpleAirReading | null] {
+  // Initialize all values
+  let id: number = Number.NaN;
+  let name: string | undefined = undefined;
+  let latitude: number | undefined = undefined;
+  let longitude: number | undefined = undefined;
+  let meanPercentDifference: number | undefined = undefined;
+  let pm25: number | undefined = undefined;
+  let humidity: number | undefined = undefined;
+  let timestamp: Date | undefined = undefined;
+
+  data.forEach((value, index) => {
+    // Check the corresponding field name to determine how to handle the value
+    switch (fieldNames[index]) {
+      case 'sensor_index':
+        if (typeof value === 'number') id = value;
```

Comment:
Would it be better to use `value instanceof number` as opposed to have a bunch of strings floating around here? Or does that only work on things in TS not from API calls?

Reply:
The internet says that `instanceof` is for objects and custom types while `typeof` is for primitives.

[Medium Article Source](https://levelup.gitconnected.com/comparing-the-javascript-typeof-and-instanceof-operators-8844f23633a)
[Stack Overflow Source](https://stackoverflow.com/questions/899574/what-is-the-difference-between-typeof-and-instanceof-and-when-should-one-be-used)